<a href="https://colab.research.google.com/github/JimenezHAnaFer/MetodosN/blob/main/Aplicaciones_problema1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Aplicaciones**

**MUERTE EN LA RESERVA**

Durante una ronda de inspección po runa reserva, dos guardabosques descubrieron el cadaver de un jabalí. Un examen preliminar permitió concluir que el animal falleció instantaneamente a causa de un balazo de un cazador furtivo. Sospechando que el cazodor debía regresar a recoger a su presa, los guardias decidieron esperarlo no lejos del lugar del crimen. Al poco tiempo aparecion dos sujetos y se dirigieron sin rodeos hacia el jabalí. Al ser detenidos, los desconocidos negaron rotundamente su participación en el delito. Aunque los guardabosques tenían pruebas indirectas de su culpabilidad, para obtner una prueba fehaciente era necesario determinar el instante exacto en que el jabalí fue muerto.

Conforme a la ley de radiación de calor, la velocidad de enfriamento de un cuerpo en el medio ambiente es propocional a la diferencia entre la temperatura del cuerpo y la temperatura del medio:





*   $\frac{dx}{dt}=-k(x-a)$




Donde **x** es la temperatura del cuerpo en el instante **t**, a es la temperatura del medio ambiente y **k** es un factor positivo de proporcionalidad, **$x_0$** es la temperatura del cuerpo en el instante **t = 0**.

a) Si en el instante en que fueron detenidos los sopechosos la temperatura *x* del cuerpo del jabalí era igual a 31°C y pasada una hora era igual a 29°C, entonces, considerando que en el instante del disparo *$x=37°C$* y *$a=29°C$*, y tomando en cuenta el instante del arresto como *$t=0$*, determinar el momento del disparo.

**Solución**



*   Tenemos los siguentes datos

>>1) $\frac{dx}{dt}=-k(x-a)$

>>2) $x(0)=31°C$ #temperatura del cuerpo del jabalí

>>3) $x(1)=29°C$ #temperatura pasada una hora $t=1$

>>4) $a=21°C$ #temperatura ambiente

>>5) t #tiempo

>>6) k #costante proporcional

>>7) $x(t)=37°C$ #temperatura en el instante del disparo en el tiempo

*   Resolvemos por EDO "variables separables:
>>$\frac{dx}{dt}=-k(x-a)$

 >>$\int\frac{dt}{x-a}=-\int kdt$

 >>$In(x-a)=-kt+C$

 >>$e^{In(x-a)}=e^{-kt+C}$

 >>$x-a=e^{-kt}e^{C}$

 >>**$x=a+Ce^{-kt}$**

*   Usamos nuestras condiciones 2,3 y 4:

 >>I) $x(0)=21+Ce^{-k(0)}=31$
 >>   $21+C=31$

 >>   $C=10$

 >>II) $x(1)=21+Ce^{-k(1)}=29$, C=10

 >>$10e^{-k}=8$

 >>$e^{-k}=\frac{8}{10}$, aplicando In en ambos lados

 >>$-k=In(\frac{8}{10})$

 >>$k=-0.223144$


 *   Obteniendo nuestra constante k, sustituimos en 7 para encontrar t (tiempo):

 >>$x(t)=21+10e^{-0.223144t}=37$

 >>$10e^{-0.223144t}=16$

 >>$e^{-0.223144t}=\frac{16}{10}$ #aplicamos In en ambos lados

 >>$-0.223144t=In(1.6)$

 >>$t=-\frac{In(1.6)}{0.223144}=-2.106279$ #redondendo en 6 decimas

 Así, entre el instante del disparo y el instante en que fueron detenidos los sospechosos transcurrieron 2 horas y 6 minutos.

b) En el caso cuando loa temperatura del aire varía con el tiempo, la ley de enfriamento del cuerpo en la ecuación diferencial lieal no homogénea:
$\frac{dx}{dt}=kx-ka(t)$.

Donde $a(t)$ es la temperatura del aire en el instante t. Hagamos algunas conjeturas: en el instante de la detención de los sospechosos la temperatura del cuerpo del jabalí era igual a 30°C; el día del acontecimiento la temperatura del aire descendió cada hora después del medio día en 1°C y era igual a 0°C en el instante del hallazgo del cadáver; finalmente pongamos que al cabo de una hora de ser detectado el animal, su temperatura bajó hasta 25°C. Tomando $t=0$ como el instante del disparo, ausmiendo que entonces la temperatura corporal del jabalí era $x_0=37°C$ y denotando con t* el instante en que los guardabosques descubrieron el cadáver, obtenemos $a(t)=t*-t$. Hallar de nuevo el momento en que murió el jabalí.

**Solución**

*   Tenemos los siguentes datos

>>1) $\frac{dx}{dt}=kx-ka(t)$

>>2) $x(0)=30°C$ #temperatura en el instante de la detención

>>3) $x(1)=25°C$ #temperatura pasada una hora $t=1$

>>4) $a(t)=t^*-t$ #temperatura ambiente

>>5) t #tiempo

>>6) k #costante proporcional

>>7) $x(t)=0°C$ #temperatura en el instante del hallazgo

*   Resolvemos EDO
  >> $\frac{dx}{dt}=kx-ka(t)$ #Tiene forma de una ED.Lineal, pero no es homogenea

  >> Se puede expresar como $x'+kx=ka(t)$ #Tiene la forma $y'+p(x)y=q(x)$

  >> Como k es una constante tenemos que $x'+p(t)kx=ka(t)$
  
  >>Ocupando la formula $xμ=\int q\mu dx$ donde $μ=e^{∫p(t)dt}$, ajustando:
  >>$xμ=∫a(t)μdx$

  >>Calculamos $μ$
  
  >> $p(t)=k$ entonces $μ=e^{∫p(t)dt}=e^{∫kdt}=e^{k∫dt}=e^{kt}$

  >>Regreando a nuestra formula:

  >>$xμ=∫a(t)μdt$ sustituyendo $μ$ tenemos:

  >>$xe^{kt}=∫a(t)e^{kt}dt$ $\to$ $xe^{kt}=∫(t^*-t)e^{kt}dt$ $\to$ $xe^{kt}=x_0-t*-\frac{1}{k}+(t^*-t+\frac{1}{k})e^{kt}$

  >> Despejando $e^{kt}$ tenemos: $x=(x_0-t^*-\frac{1}{k})e^{-kt}+t^*-t+\frac{1}{k}$

  >> Sustituyendo $x_0=37$ Tenemos: $x=(37-t^*-\frac{1}{k})e^{-kt}+t^*-t+\frac{1}{k}$......(*)

*   Tomando en cuenta 2 y 3
 >>#Tomando en cuenta que $a(t)=t^*-t=0$

 >>$x(0)=(37-t^*-\frac{1}{k})e^{-kt}+\frac{1}{k}=30$

 >>#Tomando en cuenta que $a(t)=t^*-t=1$

 >>$x(1)=(37-t^*-\frac{1}{k})e^{-k(t^*+1)}+\frac{1}{k}=26$

 >>#Apartir de esta podemos obtener la ecuación respecto a k ($t=1$ y x=25):

 >>$(30-\frac{1}{k})e^{-k}-26+\frac{1}{k}=0$

 >>#Sea $t=0$, entonces $a(t)=-t$ lo que nos lleva a la siguiente ecuación:

 >>$\frac{dx}{dt}+kx=-kt$ (con la condicion $x_0=30$ para $t=0$ despejamos x de nuestra ecuación anterior)

 >>$x=(30-\frac{1}{k})e^{-kt}-t+\frac{1}{k}$......(**)

 >>Usando nuetra ecuación (**), tomamos nuestra $x=37$ (temperatura corporal del jabalí)

 >>$(37k-1+kt)e^{kt}-30k+1=0$.....(***)

 *   Calcular nuestra constante K

 No podemos resolver algebraicamente respecto a nuestra constante k, sin embargo, podemos utilizar Métodos Numéricos para poder aproximar una raiz de nuestra constante. En este caso utilizaremos el metodo de Newton, implementado en un algoritmo en el siguiente algoritmo:

In [2]:
# Método de Newton-Raphson
# aproximemos nuestra k, que en este caso sera la variable x para terminos del codigo (k=x)

import numpy as np
import math
# INGRESO
# Nuestra función (*) la transformamos de manera que mas nos convenga
fx  = lambda x: 30*x-1+(1-26*x)*math.exp(x)
dfx = lambda x: 30-25*math.exp(x)-26*x*math.exp(x) #derivada de nuestra función

x0 = 0.5 #Se obtiene del intervalo [0,1], (a+b)/2
tolera = 0.000001

# PROCEDIMIENTO
tabla = []
tramo = abs(2*tolera)
xi = x0
while (tramo>=tolera):
    xnuevo = xi - fx(xi)/dfx(xi)
    tramo  = abs(xnuevo-xi)
    tabla.append([xi,xnuevo,tramo])
    xi = xnuevo

# convierte la lista a un arreglo.
tabla = np.array(tabla)
n = len(tabla)

# SALIDA
print(['xi', 'xnuevo', 'tramo'])
np.set_printoptions(precision = 4)
print(tabla)
print('raiz en: ', xi)
print('con error de: ',tramo)

['xi', 'xnuevo', 'tramo']
[[5.0000e-01 3.2284e-01 1.7716e-01]
 [3.2284e-01 2.2816e-01 9.4674e-02]
 [2.2816e-01 1.8850e-01 3.9665e-02]
 [1.8850e-01 1.7945e-01 9.0433e-03]
 [1.7945e-01 1.7895e-01 4.9988e-04]
 [1.7895e-01 1.7895e-01 1.5227e-06]
 [1.7895e-01 1.7895e-01 1.4124e-11]]
raiz en:  0.17895246315275215
con error de:  1.4123535674315235e-11


Para acompletar la resolución del inciso b, debemos sustituir nuestra constante $k=0.178952$ (redondeando a 6 decimales), en nuestra ecuación (***), de ese modo, podemos resolverla para la obtención de nuestra "t"

In [5]:
# Método de Newton-Raphson
# Aproximando nuestra t, que en este caso sera la variable x para el algoritmo (t=x)

import numpy as np
import math
# INGRESO
k=0.178952
fx  = lambda x: (((37*k)-1+(k*x))*math.exp(k*x))-(30*k)+1
dfx = lambda x: (37*(k**2)*math.exp(k*x))+((k**2)*x*math.exp(k*x))

x0 = -1.0
tolera = 0.000001

# PROCEDIMIENTO
tabla = []
tramo = abs(2*tolera)
xi = x0
while (tramo>=tolera):
    xnuevo = xi - fx(xi)/dfx(xi)
    tramo  = abs(xnuevo-xi)
    tabla.append([xi,xnuevo,tramo])
    xi = xnuevo

# convierte la lista a un arreglo.
tabla = np.array(tabla)
n = len(tabla)

# SALIDA
print(['xi', 'xnuevo', 'tramo'])
np.set_printoptions(precision = 4)
print(tabla)
print('raiz en: ', xi)
print('con error de: ',tramo)

['xi', 'xnuevo', 'tramo']
[[-1.0000e+00 -1.1888e+00  1.8878e-01]
 [-1.1888e+00 -1.1926e+00  3.7818e-03]
 [-1.1926e+00 -1.1926e+00  1.4802e-06]
 [-1.1926e+00 -1.1926e+00  2.2626e-13]]
raiz en:  -1.1925615459090273
con error de:  2.262634524186069e-13


Una vez aproximando nuestra t con el codigo del metodo de Newton, podemos obtener nuestra $t=-1.192562$ (redondeando a 6 cifras), por lo que el jabalí fue muerto aproximadamente 1 hora y 12 minutos antes que los guardabosques lo encontraran.